### PR TITLE
New forceios/forcedroid

### PR DESCRIPTION
**Pull request by wmathurin**
NB there are still issues being resolved - but i thought review/feedback could start
Also test_force.js has not yet been fixed

### CODE REVIEW

File: shared/createHelper.js [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+// Dependencies
+var path = require('path'),
+    SDK = require('./constants'),
+    utils = require('./utils'),
+    COLOR = require('./outputColors');
+
+//
+// Helper for native application creation
+//
+function createNativeApp(config) {
```
Comment (wmathurin) [inline shared/createHelper.js]: That's the function that does native app creation (for ios and android).

File: shared/createHelper.js [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    utils.removeFile(path.join(config.projectDir, 'template.js'));
+
+    // Done
+    return prepareResult;
+}
+
+//
+// Helper for hybrid application creation
+//
+function createHybridApp(config) {
```
Comment (wmathurin) [inline shared/createHelper.js]: That's the function that does hybrid app creation (for ios and android).

File: android/forcedroid.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+// Dependencies
+var SDK = require('./shared/constants'),
```
Comment (bhariharan) [inline android/forcedroid.js]: Wow that's really short!

File: test.sh [hunk 4]
```diff
@@ -0,0 +1,10 @@
+node ios/forceios.js create --apptype=native --appname=App1 --packagename=com.acme.app1 --organization=Acme1 --outputdir=app1 --templaterepourl
```
Comment (wmathurin) [inline test.sh]: This file will go soon (once test_force.js is fixed).
For now I used it for local manual testing.

File: shared/constants.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+module.exports = {
```
Comment (wmathurin) [inline shared/constants.js]: All the constants live here.

File: shared/constants.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    version: '5.0.0',
+
+    cordova: {
+        pluginRepoUrl: '/Users/wmathurin/Development/github/wmathurin/SalesforceMobileSDK-CordovaPlugin',
```
Comment (wmathurin) [inline shared/constants.js]: I will fix that soon.

File: shared/constants.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        }
+    },
+
+    appTypes: {
+        ios: ['native', 'native_swift', 'react_native', 'hybrid_local', 'hybrid_remote'],
+        android: ['native', 'react_native', 'hybrid_local', 'hybrid_remote']
+    },
+
+    templates: {
+        repoUrl: 'https://github.com/wmathurin/SalesforceMobileSDK-Templates',
```
Comment (wmathurin) [inline shared/constants.js]: I'll fix that soon.

File: shared/constants.js (hunk 6 above)
Comment (bhariharan) [inline shared/constants.js]: Change to `forcedotcom` repo URL.

File: shared/constants.js (hunk 7 above)
Comment (bhariharan) [inline shared/constants.js]: Change to `forcedotcom`.

File: shared/pack.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env node
```
Comment (wmathurin) [inline shared/pack.js]: That's the script to run to generate the forceios-xxx.tgz and forcedroid-xxx.tgz package.
To generate both packages (in the current working directory)
./shared/pack.js --os=ios,android

**Comment by wmathurin** [comment]
Current state of affairs:

* iOS

| App Type      | Generate | Build | Compile | Run                               |
| ------------- | -------- | ----- | ------- | --------------------------------  |
| native        | yes      | yes   | yes     | yes (but not on iOS 10 simulator) |
| native_swift  | yes      | yes   | yes     | yes                               |
| react_native  | yes      | yes   | yes     | yes                               |
| hybrid_local  | yes      | yes   | no      | no                                |
| hybrid_remote | yes      | yes   | no      | no                                |


* Android

| App Type      | Generate | Build | Compile | Run                   |
| ------------- | -------- | ----- | ------- | --------------------- |
| native        | yes      | yes   | yes     | yes                   |
| react_native  | yes      | yes   | yes     | no (crash at startup) |
| hybrid_local  | yes      | yes   | no      | no                    |
| hybrid_remote | yes      | yes   | no      | no                    |

**Comment by wmathurin** [comment]
Please have a look at shared/pack.js (the script to npm pack forceios and forcedroid) and test/test_force.js.

File: test/test_force.js [hunk 9]
```diff
@@ -111,263 +104,119 @@ function main(args) {
...
-}
-
-//
 // Helper to validate version
 // 
 function validateVersion(version) {
     if (version.match(/\d\.\d\.\d/) == null) {
-            log('Invalid version: ' + version, COLOR.red);
-            process.exit(1);
+        utilslog('Invalid version: ' + version, COLOR.red);
```
Comment (khawkins) [inline test/test_force.js]: Is this a typo?

Comment (wmathurin) [inline test/test_force.js]: Indeed! Actually the whole method can go. The version is now pulled from shared/constants.js.

**Comment by wmathurin** [comment]
@khawkins @bhariharan @rwhitley 

What do you think of https://github.com/forcedotcom/SalesforceMobileSDK-Package/pull/3/commits/da502ab8262c0ff87517f69eafe5ff3a1929c947 ?

Basically 'forceios/droid create' asks you the bare minimum.
Advanced options (output directory, template repo url, plugin repo url) are available through 'forceios/droid create2'. 

Most people, I expect will use forceios/droid create.

```shell
forceios create --apptype=native --appname=app1 --packagename=com.acme.app1 --organization=Acme1

forcedroid create --apptype=native --appname=app1 --packagename=com.acme.app1 --organization=Acme1

``` 

(Maybe calling create2 is not very inspired, so recommendations welcome).

File: shared/configHelper.js [hunk 10]
```diff
@@ -44,34 +44,40 @@ function readConfig(args, toolName, toolVersion, appTypes, handler) {
...
+    log('    --pluginrepourl=<Cordova plugin URL> (Leave empty for version ' + toolVersion + ' of mobile sdk plugin.)]', COLOR.magenta);
     log('\n OR \n', COLOR.cyan);
     log(toolName + ' version', COLOR.magenta);
 }
 
 //
 // Processor list for 'create' command
 //
-function createArgsProcessorList(appTypes) {
+function createArgsProcessorList(appTypes, extraArgs) {
```
Comment (khawkins) [inline shared/configHelper.js]: Can this be more boolean-ish?  E.g. `presentExtraArgs`?

**Comment by bhariharan** [comment]
@wmathurin How about `createWithConfig` instead of `create2`? I like the idea.

**Comment by wmathurin** [comment]
Let's say you don't want to specify the template repo url, then with optional argument, you have to pass in --templaterepourl, otherwise you end up being prompted. So it's a bit painful, the above creation examples look like:

```
forceios create --apptype=native --appname=app1 --packagename=com.acme.app1 --organization=Acme1 --outputdir --templaterepourl --pluginrepourl
```

**Comment by khawkins** [comment]
Makes sense.  Seems okay to me.

**Comment by wmathurin** [comment]
At this point, what's left is really bug fixes. So I'm going to merge this one in, and then more PRs with the fixes (that will also allow others to join in on the effort).

**Comment by khawkins** [comment]
👍

**Pull request merged**